---
title: "Untitled"
output: html_document
date: "2023-03-25"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r cars}


library(tidyr)
library(dplyr)

# set the path to the directory containing the files
path <- "C:/Users/Briana/Desktop/Uni/Briana Project Stuff/getVEPRunning/ProcessedVEPInsertions"

# list all the text files in the directory
files <- list.files(path, pattern = "\\.txt$", full.names = TRUE)

# read and process each file, and combine the data into one table
combined_df <- lapply(files, function(file) {
  # read the file
  vcf <- read.table(file, header = FALSE, comment.char = "#", sep = "\t")
  
  # split the 8th column into multiple columns
  new_df <- separate(vcf, 8, into = c("1", "2", "Consequence", "Gene", "5", "6", "7", "8", "9", "10", "11", "12", "13", "14", "15", "16", "17", "Existing Variant", "19", "20", "21", "22", "23", "24", "25", "26", "27", "28", "29", "30", "31", "32", "33", "34", "35", "36", "37", "38", "39", "40", "41", "42"), sep = "\\|")
  
  # add the file name as another column
  new_df$file_name <- file
  
  # return the processed data
  new_df
})

# combine all the data frames into one and add the file name as a column
combined_df <- bind_rows(combined_df) %>% 
  mutate(file_name = basename(file_name)) %>% 
  select(file_name, everything())




```

```{r create split DB}

df_split <- separate(combined_df, '1', into = c("EVENT", "PAIRID", "SVLEN","SVTYPE","CSQ"), sep = ";")
df_split$SVLEN <- gsub("SVLEN=", "", df_split$SVLEN)
df_split$SVLEN <- as.integer(df_split$SVLEN)



```

```{r HIGH or MODERATE con}

library(ggplot2)

# Subset data to include only variants that start with "COS"
combined_df_sub <- subset(combined_df, grepl("^COS", `Existing Variant`) & (`Consequence` == "HIGH" | `Consequence` == "MODERATE"))

# Aggregate data by gene and variant and count the number of occurrences
gene_counts <- aggregate(`Existing Variant` ~ Gene, data = combined_df_sub, FUN = function(x) length(x))

# Order the genes by count of variants
gene_counts_sorted <- gene_counts[order(gene_counts$`Existing Variant`, decreasing = TRUE), ]

# Create bar plot
highmodcos <- ggplot(gene_counts_sorted, aes(x = reorder(Gene, `Existing Variant`), y = `Existing Variant`, fill = `Existing Variant`)) +
  geom_bar(stat = "identity") +
  labs(fill = "Consequence\n by Count",x = "Gene", y = "HIGH and MODERATE Consequence count") +  theme_light() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        plot.title = element_text(hjust = 0.5)) + 
  ggtitle("VEP determined HIGH or MODERATE consequence MINTIE deletions with\n associated COSMIC entry in TCGA TARGET AML cohort")

filepath <- "/Users/Briana/Desktop/Uni/Briana Project Stuff/getVEPRunning/ProcessedDeletions/HIGHMODCOS.png" 
png(filepath, width = 1300, height = 1100)
plot(highmodcos)
dev.off()




```

```{r fig.height = 20, fig.width = 10}
library(ggplot2) 
library(ggthemes)

# Subset rows where consequence is HIGH or MODERATE
df_subset <- subset(combined_df, Consequence %in% c("HIGH", "MODERATE"))

# Aggregate data by gene and consequence and count the number of occurrences
gene_counts <- aggregate(Consequence ~ Gene, data = df_subset, FUN = function(x) length(x))

# Order the genes by count of consequences and select top 20
top_20_genes <- gene_counts[order(gene_counts$Consequence, decreasing = TRUE), ][1:50,]

# Create bar plot
topfifty <- ggplot(top_20_genes, aes(x = reorder(Gene, Consequence), y = Consequence, fill = Consequence)) +
  geom_bar(stat = "identity") +
  labs(fill = "Consequence\n by Count",x = "Gene", y = "HIGH and MODERATE Consequence Combined Count ", title = "VEP determined top 50 genes by count associated with MODERATE and HIGH consequence MINTIE deletions\n in TCGA TARGET AML cohort", ) +
  theme_light() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),  plot.title = element_text(hjust = 0.5))

filepath <- "/Users/Briana/Desktop/Uni/Briana Project Stuff/getVEPRunning/ProcessedDeletions/HIGHMODTOP50.png"  
png(filepath, width = 1300, height = 1100)
plot(topfifty)
dev.off()


```



```{r Insertion length distribution}
library(ggplot2)

distribution_plot <- ggplot(df_split, aes(x = SVLEN)) +
  geom_histogram(alpha = 0.5, fill = "Lightgreen", color = "Darkgreen", binwidth = 5) +
  labs(y = "Count", x = "Deletion length (bp)",title = "Length distribution of MINTIE deletions\n in TCGA Target AML Cohort") +
  theme_light() +
  guides(x = guide_axis(angle = 90))

modifieddistplot <- distribution_plot + 
  theme(plot.title = element_text(size = 20,hjust = 0.5), 
        axis.title.x = element_text(size = 15),
        axis.title.y = element_text(size = 15), 
        legend.position = "bottom",
        )


filepath <- "/Users/Briana/Desktop/Uni/Briana Project Stuff/getVEPRunning/ProcessedDeletions/DeletionLengthDistribution.png" 
png(filepath, width = 1300, height = 1100)
plot(modifieddistplot)
dev.off()


```




```{r fig.height = 10, fig.width = 5}

library(dplyr)

genes <- df_split %>% 
  filter(Consequence %in% c("MODERATE", "HIGH")) %>% 
  group_by(Gene) %>% 
  summarize(count = n()) %>% 
  arrange(desc(count)) %>% 
  head(26) %>% 
  pull(Gene)


gene_df <- data.frame(genes)
write.csv(gene_df,"/Users/Briana/Desktop/Uni/Briana Project Stuff/getVEPRunning/ProcessedDeletions/TopGenes.png/Users/Briana/Desktop/Uni/Briana Project Stuff/getVEPRunning/ProcessedDeletions/DeletionLengthDistribution.png", row.names = FALSE)



```

```{r}
library(dplyr)

# Filter for genes with COSV entries and moderate/high consequences
genes <- df_split %>%
  filter(grepl("COSV", `Existing Variant`) & Consequence %in% c("MODERATE", "HIGH")) %>%
  group_by(Gene, `Existing Variant`) %>%
  summarize(count = n()) %>%
  ungroup()

# Write the list of genes to a CSV file
write.csv(genes, "/Users/Briana/Desktop/Uni/Briana Project Stuff/getVEPRunning/ProcessedDeletions/cosmic.csv", row.names = FALSE)



```

```{r fig.height = 60, fig.width = 30}
library(ggplot2)

# Subset the data to include HIGH, MODERATE, and LOW consequence variants
df_filtered <- df_split[df_split$Consequence %in% c("HIGH", "MODERATE"),]

# Aggregate the data by GenomicImpact
agg_data <- aggregate(file_name ~ GenomicImpact, data = df_filtered, FUN = length)

# Create the bar chart
genomicimpactnomodifier <- ggplot(agg_data, aes(x = GenomicImpact, y = file_name, fill = GenomicImpact)) + theme_light() +
  geom_bar(stat = "identity") +
  scale_fill_brewer(palette = "Set2") +
  labs(title = "HIGH and MODERATE MINTIE deletion\n genomic impact distribution in TCGA TARGET AML",
       x = "Genomic Impact",
       y = "Count") + 
  theme(plot.title = element_text(hjust = 0.5, size = 20),
        axis.text.x = element_text(angle = 45, hjust = 1, size =15),
        axis.title.x =element_text(size =20),
        axis.title.y = element_text(size = 20),
        axis.text = element_text(size = 15),
        legend.text = element_text(size = 14),
        legend.title = element_text(size = 14))




filepath <- "/Users/Briana/Desktop/Uni/Briana Project Stuff/getVEPRunning/ProcessedDeletions/GenomicImpactNoModifier.png" 
png(filepath, width = 1300, height = 1100)
plot(genomicimpactnomodifier)
dev.off()





```

```{r fig.height = 50, fig.width = 25}
# Create a new column to store the updated existing variant entries
df_split$UpdatedExistingVariant <- ""

# Loop through each row in the df_split data frame
for(i in 1:nrow(df_split)) {
  
  # Check if the existing variant contains "COSV" or "RS" or both
  has_cosv <- grepl("COSV", df_split[i, "Existing Variant"])
  has_rs <- grepl("rs", df_split[i, "Existing Variant"])
  
  # Assign a value to the UpdatedExistingVariant column based on the presence of "COSV" and "RS"
  if(has_cosv & has_rs) {
    df_split[i, "UpdatedExistingVariant"] <- "COS&rs"
  } else if(has_cosv) {
    df_split[i, "UpdatedExistingVariant"] <- "COSV"
  } else if(has_rs) {
    df_split[i, "UpdatedExistingVariant"] <- "rs"
  } else if(df_split[i, "Existing Variant"] == "N/A") {
    df_split[i, "UpdatedExistingVariant"] <- "NONE"
  } else {
    df_split[i, "UpdatedExistingVariant"] <- "Other"
  }
}

# Create a new data frame with the ExistingVariant, UpdatedExistingVariant, and Consequence columns
df_existing <- df_split[, c("Existing Variant", "UpdatedExistingVariant", "Consequence", "GenomicImpact")]



```


```{r fig.height = 100, fig.width = 50}
library(ggplot2)
library(ggplot2)

df_existing <- df_existing[df_existing$Consequence != "MODIFIER",]

library(ggplot2)

# Count the number of variants for each value in the UpdatedExistingVariant column
agg_data <- aggregate(df_existing$Consequence, 
                      by = list(df_existing$UpdatedExistingVariant, df_existing$GenomicImpact), 
                      FUN = length)
names(agg_data) <- c("UpdatedExistingVariant", "GenomicImpact", "Count")

# Create the bar chart
databaseentries <- ggplot(agg_data, aes(x = UpdatedExistingVariant, y = Count, fill = GenomicImpact)) +
  geom_bar(stat = "identity") +
  scale_fill_brewer(palette = "Set2") + theme_light() +
  labs(title = "VEP MODERATE and HIGH Consequence MINTIE deletions by associated database\n in TCGA TARGET AML",
       x = "Existing Variant",
       y = "Count") +
  theme(plot.title = element_text(hjust = 0.5, size = 25),
        axis.text.x = element_text(angle = 45, hjust = 1, size =15),
        axis.title.x =element_text(size =20),
        axis.title.y = element_text(size = 20),
        axis.text = element_text(size = 15),
        legend.text = element_text(size = 14),
        legend.title = element_text(size = 14))

filepath <- "/Users/Briana/Desktop/Uni/Briana Project Stuff/getVEPRunning/ProcessedDeletions/databaseentries.png" 
png(filepath, width = 1300, height = 1100)
plot(databaseentries)
dev.off()


```


```{r fig.height = 250, fig.width = 175}


# Aggregate the data by GenomicImpact
agg_data <- aggregate(file_name ~ Consequence, data = df_split, FUN = length)

# Create the bar chart
genomicimpactnomodifier <- ggplot(agg_data, aes(x = Consequence, y = file_name)) + theme_light() +
  geom_bar(stat = "identity") +
  scale_fill_brewer(palette = "Set2") +
  labs(title = "Consequence distribution for MINTIE insertions in TCGA Target AML cohort",
       x = "VEP determined Genomic Impact",
       y = "Deletion Count") + 
  theme(plot.title = element_text(hjust = 0.5, size = 30),
        axis.text.x = element_text(angle = 45, hjust = 1, size =15),
        axis.title.x =element_text(size =20),
        axis.title.y = element_text(size = 20),
        axis.text = element_text(size = 15),
        legend.text = element_text(size = 14),
        legend.title = element_text(size = 14))




filepath <- "/Users/Briana/Desktop/Uni/Briana Project Stuff/getVEPRunning/ProcessedVEPInsertions/consequencedistribution.png" 
png(filepath, width = 1300, height = 1100)
plot(genomicimpactnomodifier)
dev.off()








```